File: src/python/yx/primitives/operation_result.py
"""
Operation result wrapper for error handling

Provides a Result type similar to Swift's Result<T, Error>.
"""

from typing import Generic, TypeVar, Callable, Union
from dataclasses import dataclass

T = TypeVar('T')
E = TypeVar('E', bound=Exception)


@dataclass
class Success(Generic[T]):
    """Represents a successful operation result"""
    value: T

    def is_success(self) -> bool:
        return True

    def is_failure(self) -> bool:
        return False

    def get_value(self) -> T:
        return self.value

    def get_error(self) -> None:
        return None


@dataclass
class Failure(Generic[E]):
    """Represents a failed operation result"""
    error: E

    def is_success(self) -> bool:
        return False

    def is_failure(self) -> bool:
        return True

    def get_value(self) -> None:
        return None

    def get_error(self) -> E:
        return self.error


# Type alias for convenience
OperationResult = Union[Success[T], Failure[E]]


def result_map(result: OperationResult, func: Callable[[T], T]) -> OperationResult:
    """
    Map a function over a successful result.

    Args:
        result: Operation result
        func: Function to apply to success value

    Returns:
        OperationResult: New result with mapped value, or original failure
    """
    if isinstance(result, Success):
        try:
            return Success(func(result.value))
        except Exception as e:
            return Failure(e)
    return result


def result_flat_map(result: OperationResult, func: Callable[[T], OperationResult]) -> OperationResult:
    """
    Flat map a function over a successful result.

    Args:
        result: Operation result
        func: Function that returns an OperationResult

    Returns:
        OperationResult: Result from func, or original failure
    """
    if isinstance(result, Success):
        try:
            return func(result.value)
        except Exception as e:
            return Failure(e)
    return result

File: Sources/Primitives/OperationResult.swift
import Foundation
// filename: OperationResult.swift


/// Result type for operations that can fail with multiple error types
public enum OperationResult<Success> {
    case success(Success)
    case failure(Error)

    /// Unwrap success value or throw error
    public func get() throws -> Success {
        switch self {
        case .success(let value):
            return value
        case .failure(let error):
            throw error
        }
    }

    /// Map successful value
    public func map<T>(_ transform: (Success) -> T) -> OperationResult<T> {
        switch self {
        case .success(let value):
            return .success(transform(value))
        case .failure(let error):
            return .failure(error)
        }
    }

    /// Returns true if this is a success
    public var isSuccess: Bool {
        if case .success = self {
            return true
        }
        return false
    }

    /// Returns true if this is a failure
    public var isFailure: Bool {
        !isSuccess
    }
}
